wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+
+To increase the security strength of our solution, we can combine this technique with some of the following, to make even AntMiner impractical for double OTP, and to make GPU attack exceedingly slow and uneconomical even for single OTP.
+
+
+#### II. Complex Hash Function
+
+We can improve step (2) of **GenLeaf** by replacing the SHA256 hash function with a more complex alternative. Ideally, the alternative should be easy to compute on a user's device (using standard CPU and RAM) but much "harder" on devices for massive parallel computing (e.g. GPU, FPGA, ASIC). A natural starting point is scrypt, which can be configured to use an arbitrarily large amount of memory that is required to produce the hash result. In particular, scrypt is used by multiple cryptocurrency (such as Litecoin, Doge, and others) in their mining algorithms with the initial intent to make mining less advantageous on GPU, and to be ASIC-resilient. The design was effective initially<sup>[1](#f1)</sup>, but ultimately failed to achieve ASIC resilience<sup>[2](#f2)</sup>, because over the last few years it became considerably easier to fit reasonably sized memory chips on ASIC, and the choice of the security parameters did not consider such advancement. Nonetheless, based on our own benchmark result using NVIDIA Tesla K80 GPUs, by replacing the hash function to scrypt (even under very low memory settings), it would roughly increase 100-1000 the amount of time the attacker would need to complete the bruteforce for a client secured with a single OTP and the Controlled Randomness technique in (1).
+
+After comparing several alternatives, we believe [argon2](https://github.com/P-H-C/phc-winner-argon2) is the optimal choice to replace step (2) of **GenLeaf** in the next version of ONE Wallet. Similar to scrypt, it can be configured to use a large size of memory and impose an arbitary workload to slow down the computation. Aside for its multiple advantages over other candidates (which are discussed at length in their Github repository, the password-hashing competition, and their paper), here are the other factors we have considered:
+
+- Argon2 has a [browser-based implementation](https://github.com/antelle/argon2-browser) based on WebAssembly. It offers satisfactory performance under reasonably low memory settings, and can complete within a reasonable amount of time for 10<sup>6</sup> repeatitions. Therefore, it would not slow down the setup process
+- Based on our knowledge, no one has built an ASIC miner for computing Argon2 hashes at this time. 
```

Comment:
From @ivan-homoliak-sutd  : 

> This might change in the future, especially if OneWallet will be successfull :) What about to employ verifiable delay functions (VDFs), such as recursice hash computations (e.g., https://en.wikipedia.org/wiki/PBKDF2) instead of scrypt or on top of it?

Reply:
This sounds promising